cluster-agent/controllers/managed-gitops/eventloop/operation_event_loop.go

```diff
@@ -1222,3 +1331,122 @@ func generateExpectedClusterSecret(ctx context.Context, application db.Applicati
 	return managedEnvironmentSecret, deleteSecret_false, nil
 
 }
+
+func buildAppProject(ctx context.Context, dbOperation db.Operation, opConfig operationConfig, log logr.Logger) (*appv1.AppProject, error) {
+
+	// Create AppProject resource before creating Argo CD Application CR
+
+	var appProjectRepositories []db.AppProjectRepository
+	if err := opConfig.dbQueries.ListAppProjectRepositoryByClusterUserId(ctx, dbOperation.Operation_owner_user_id, &appProjectRepositories); err != nil {
+		log.Error(err, "unable to list AppProjectRepositories based on cluster user id")
+		return nil, err
+	}
+
+	repoURLs := []string{} // Create a new slice to store RepoURLs
+	// Iterate over the appProjectRepositories and append RepoURLs to the repoURLs slice
+	for _, repo := range appProjectRepositories {
+		repoURLs = append(repoURLs, repo.RepoURL)
+	}
+
+	var clusterSecretNames []string
+
+	var appProjectManagedEnvs []db.AppProjectManagedEnvironment
+	if err := opConfig.dbQueries.ListAppProjectManagedEnvironmentByClusterUserId(ctx, dbOperation.Operation_owner_user_id, &appProjectManagedEnvs); err != nil {
+		log.Error(err, "unable to list appProjectManagedEnvs by cluster user id")
+		return nil, err
+	}
+
+	for _, appProjectManagedEnv := range appProjectManagedEnvs {
+
+		managedEnv := db.ManagedEnvironment{
+			Managedenvironment_id: appProjectManagedEnv.Managed_environment_id,
+		}
+
+		if err := opConfig.dbQueries.GetManagedEnvironmentById(ctx, &managedEnv); err != nil {
+			log.Error(err, "unable to retrieve managedEnv by id")
+			return nil, err
+		}
+
+		clusterSecretNames = append(clusterSecretNames, argosharedutil.GenerateArgoCDClusterSecretName(managedEnv))
+	}
+
+	var destinations []appv1.ApplicationDestination
+	for _, clusterSecretName := range clusterSecretNames {
+		destinations = append(destinations, appv1.ApplicationDestination{
+			Name:      clusterSecretName,
+			Namespace: "*",
+		})
+	}
+
+	// Make sure we also add the local cluster
+	destinations = append(destinations, appv1.ApplicationDestination{
+		Name:      ArgoCDDefaultDestinationInCluster,
+		Namespace: "*",
+	})
+
+	appProject := &appv1.AppProject{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: appProjectPrefix + dbOperation.Operation_owner_user_id,
+			Annotations: map[string]string{
+				"username": dbOperation.Operation_owner_user_id,
+			},
+			Namespace: opConfig.argoCDNamespace.Name,
+		},
+		Spec: appv1.AppProjectSpec{
+			SourceRepos:  repoURLs,
+			Destinations: destinations,
+		},
+	}
+
+	return appProject, nil
+
+}
+
+func appProjectEqual(existingAppProject, generatedAppProject *appv1.AppProject) bool {
+
+	if existingAppProject == nil || generatedAppProject == nil {
+		return false
+	}
+
+	// Check if length of existingAppProject.Spec.SourceRepos is equal to generatedAppProject.Spec.SourceRepos
+	if len(existingAppProject.Spec.SourceRepos) != len(generatedAppProject.Spec.SourceRepos) {
+		return false
```

Comment:
Thanks for adding the unit tests for `appProjectEqual`!  One final request, then this PR is good to go: 
- Since this function is very important to the overall correctness of the app project logic, it would be good to make sure the unit tests are test 100% of the branches/blocks of this function (that is, ensure that all lines are covered)
- We the unit test at present, it looks like we currently don't have coverage for most of the inside of the if{} blocks
    -  For example, for line 1413 here, we could add a test that verifies that false is returned if the length of the source repos is different.
    - Likewise for other cases where we are returning `false` in this function.